Create

oxinabox opened the pull request:
Checklist from https://github.com/JuliaCollections/AxisArraysFuture/issues/1#issuecomment-482077891

- [x] Wrapper Type, that wraps an `AbstractArray` giving names to the indexes. The index named can go in the type params
 - [x] Mapping function: `name2dim(::NamedDimsArray, ::Symbol)::Int` that will take an array with named dimentsions and dimention name, and return its dimension index.
 - [x] Overloads for common functions that take a `dims` argument. Like `sum` and `mean`
       - that return `NamedDimArrays`
 - [x] kwarg style indexing https://github.com/JuliaLang/julia/issues/25631
 - [x] Optional `AbstractArray` methods [see docs](https://docs.julialang.org/en/v1/manual/interfaces/index.html#man-interface-array-1)

oxinabox commented on src/wrapper_array.jl:
```diff
@@ -0,0 +1,81 @@
+    inds = map(dim_names(A)) do name
+        get(named_inds, name, :)  # default to slicing
+    end
+end
```
Alternative from @iamed2 

Something that does
```
defaults = (a=:, b=:, c=:)  # In the right order
Tuple(merge(defaults, named_inds))
```

So we would need a function that generates that default.

@iamed2  says this would give compile time sorting,
though I am not sure it would with the Error we throw.
Compile time **hates** things that can throw exceptions.

iamed2 replied:
If the error means bail out, compile time is fine with it. I believe it's only a problem when you're catching exceptions or when the exception can change the output.

oxinabox replied:
I will have to recheck tgat

oxinabox replied:
Seems right

```
julia> @code_llvm (()->NamedDims.order_named_inds((:a,)))()

;  @ none:1 within `#41'
define nonnull %jl_value_t addrspace(10)* @"japi1_#41_14567"(%jl_value_t addrspace(10)*, %jl_value_t addrspace(10)**, i32) #0 {
top:
  %3 = alloca %jl_value_t addrspace(10)**, align 8
  store volatile %jl_value_t addrspace(10)** %1, %jl_value_t addrspace(10)*** %3, align 8
  ret %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 5019490584 to %jl_value_t*) to %jl_value_t addrspace(10)*)
}

julia> @code_llvm (()->NamedDims.order_named_inds((:a,), b=2))()

;  @ none:1 within `#43'
; Function Attrs: noreturn
define nonnull %jl_value_t addrspace(10)* @"japi1_#43_14568"(%jl_value_t addrspace(10)*, %jl_value_t addrspace(10)**, i32) #0 {
top:
  %3 = alloca %jl_value_t addrspace(10)**, align 8
  store volatile %jl_value_t addrspace(10)** %1, %jl_value_t addrspace(10)*** %3, align 8
  %4 = call nonnull %jl_value_t addrspace(10)* @"julia_#order_named_inds_14569"({ i64 } addrspace(11)* nocapture readonly addrspacecast ({ i64 }* @0 to { i64 } addrspace(11)*), %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 4595989792 to %jl_value_t*) to %jl_value_t addrspace(10)*))
  call void @llvm.trap()
  unreachable
}
```

oxinabox commented:
probably want to implement a bunch of the optional `AbstractArray` stuff

iamed2 commented on src/name2dim.jl:
```diff
@@ -0,0 +1,47 @@
+
+###################################
+
+# We change the name into a `Val` because that will trigger constant propergation
```
This doesn't need to be a val for constant propagation, just for dispatch

oxinabox replied:
Correct comment was incorrect notes from earlier experiments in making this compile time

iamed2 commented on src/name2dim.jl:
```diff
@@ -0,0 +1,47 @@
+
+# We change the name into a `Val` because that will trigger constant propergation
+# and allow this to resolve at compile time when the `name` is a constant
+name2dim(namemap, name) = _name2dim(namemap, Val(name))
```
If you have a trivial NamedTuple where names point to their position in the tuple:
```julia
dims = (foo = 1, bar = 2)
```
Then this function will do what `name2dim` does at compile time if possible:
```
name2dim(dims_nt, name) = first(merge(NamedTuple{(name,), Tuple{Int}}((0,)), dims_nt))
```

oxinabox commented:
Julia 1.1 is noticable better at constant propergation such that things happen at compile time.

nicoleepp commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+"""
+    name2dim(dimnames, [name])
+
+For `dimnames` being a tuple of dimnames (symbols) for dimenensions.
```
```suggestion
For `dimnames` being a tuple of dimnames (symbols) for dimensions.
```

nicoleepp commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+
+For `dimnames` being a tuple of dimnames (symbols) for dimenensions.
+If called with just the tuple,
+returns a named tuple, with each name maps to a dimension.
```
```suggestion
returns a named tuple, with each name mapped to a dimension.
```

nicoleepp commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+end
+
+function name2dim(dimnames::Tuple, dim::Union{Integer, Colon})
+    # This is the fallback that allows `NamedDimsArray`'s to be have dimenstions
```
```suggestion
    # This is the fallback that allows `NamedDimsArray`'s to have dimensions
```

rofinn replied:
```suggestion
    # This is the fallback that allows `NamedDimsArray`'s to be have dimensions
```

nicoleepp commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+
+"""
+    default_inds(dimnames::Tuple)
+This is the defult value for all indexing expressions using the given dimnames.
```
```suggestion
This is the default value for all indexing expressions using the given dimnames.
```

iamed2 commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+e.g. `name2dim((:a, :b), :b) == 2`
+If that `name` is not found then `0` is returned.
+"""
+function name2dim(dimnames::Tuple)
```
```suggestion
function name2dim(dimnames::NTuple{N, Symbol}) where N
```
And then you could get rid of the second line of the function

iamed2 replied:
Also applies to `default_inds`

iamed2 commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+function order_named_inds(dimnames::Tuple; named_inds...)
+    # Note: This code is runnable at compile time if input is a constant
+    # If modified, make sure to recheck that it still can run at compile time
+    keys(named_inds) ⊆ dimnames || throw(
```
An alternative to this would be checking to see that `length(full_named_inds) == length(dimnames)`

rofinn commented on src/NamedDims.jl:
```diff
@@ -1,5 +1,11 @@
+using Statistics
 
-greet() = print("Hello World!")
+export NamedDimsArray, name2dim, dim_names
```
Seems weird to only partially follow the `_` convention.

nickrobinson251 replied:
i prefer `a2b` naming only for `Dict`s but I don't have a good alternative and it's only an internal function

oxinabox replied:
so `name_to_dim`?
I am ok this that.

oxinabox replied:
`get_dim`?
`dim`?

nickrobinson251 replied:
i quite to like `dim(names)` and `dim(names, name)` ...but maybe it's scarily short?

oxinabox replied:
We have `axes` it is only slightly shorter.

oxinabox replied:
Can't do `dims`,
it is a kwarg on functions that want  to call this.

e.g.  
`sum(xs::NamedDimArray; dims)=sum(parent(xs), dims(xs, dims))`
does  not work; and
`sum(xs::NamedDimArray; dims)=sum(parent(xs), NamedDimArray.dims(xs, dims))`
is ugly.

rofinn commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+"""
+    name2dim(dimnames, [name])
+
+For `dimnames` being a tuple of dimnames (symbols) for dimenensions.
```
I might go with `name2dim(dn, [name])` to avoid having a function called `dim_names` and a variable called `dimnames`.

oxinabox replied:
Perhaps `dim_labels`?

rofinn replied:
Sure... or maybe `name2dim(names, [labels])`?

oxinabox replied:
now that it is called `dim` this is resolved

iamed2 commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+Dimensions indexed with scalars are dropped
+"""
+@generated function determine_remaining_dim(dimnames::Tuple, inds)
+    # TODO: This still allocates once, and it shouldn't have to
```
I'll note the equivalent code with constants also allocates:
```julia
       function gend()
           v1 = getfield((:a, :b, :c), 1)::Symbol
           v2 = getfield((:a, :b, :c), 3)::Symbol
           Core.tuple(v1, v2)::Tuple{Symbol, Symbol}
       end
```

oxinabox replied:
Yes

iamed2 commented on src/name_core.jl:
```diff
@@ -0,0 +1,97 @@
+determine which are not dropped.
+Dimensions indexed with scalars are dropped
+"""
+@generated function determine_remaining_dim(dimnames::Tuple, inds)
```
```suggestion
@generated function determine_remaining_dim(dimnames::Tuple, inds::Tuple)
```

iamed2 commented on src/wrapper_array.jl:
```diff
@@ -0,0 +1,115 @@
+# `L` is for labels, it should be a `Tuple` of `Symbol`s
```
```suggestion
# `L` is for labels, it should be an `NTuple{N, Symbol}`
```
To clarify that it should be N items long

iamed2 commented on src/wrapper_array.jl:
```diff
@@ -0,0 +1,115 @@
+Base.axes(a::NamedDimsArray, dim) = axes(parent(a), name2dim(a, dim))
+
+
+function Base.similar(a::NamedDimsArray{L}, args...) where L
```
This will only work when the returned array from `similar` has the same number of dimensions. Maybe the type signature can reflect this?

oxinabox replied:
I don't think so.
`similar` has *so many* overloads.
And we want to leave them the args alone.

iamed2 replied:
Based on looking at all the overloads in stdlib/Base, I think it's pretty easy to separate it into two categories: one with dims, and one without. And it would be simple to make this only apply without dims:
```suggestion
function Base.similar(a::NamedDimsArray{L}, args::Type...) where L
```

oxinabox replied:
Why the `...`s?
This can't match anything that is used except the type form.
Which is probably ok,
but in that case `...` isn't needed

iamed2 replied:
Some SparseArrays methods take two types. 

It appears the only thing that is ever used that isn't types is dims.

oxinabox replied:
Ok, cool

rofinn commented on src/wrapper_array.jl:
```diff
@@ -0,0 +1,115 @@
+# `L` is for labels, it should be a `Tuple` of `Symbol`s
+struct NamedDimsArray{L, T, N, A<:AbstractArray{T,N}} <: AbstractArray{T,N}
```
Docstring for the type or constructor?

rofinn commented on src/wrapper_array.jl:
```diff
@@ -0,0 +1,115 @@
+
+Returns a tuple of containing the names of all the dimensions of the array `A`.
+"""
+dim_names(::Type{<:NamedDimsArray{L}}) where L = L
```
I'd also be fine with just calling this `names` and not exporting it...?

nickrobinson251 replied:
alternatives: `keys` (like NamedTuples, etc) or `Base.names` 
not sure i prefer either of them to just `names`

oxinabox replied:
I like `names`

oxinabox replied:
I like `names`, goes well with `dims`

oxinabox commented:
> I feel like you would want to return a NamedDimsArray from operations on a NamedDimsArray

Ah yes, indeed I do.
Did not think `sum` etc through fully.

Looks like I will have to implement more functions for working out dimensions, after operations
but that should make #2 #3 #4 easier at least

oxinabox commented:
Added dropdims, because I implemented the code for it by mistake (I thought summing over a dimension dropped that dimension, but apparently not (did it used to?))

This is ready for another round of review.

nickrobinson251 commented:
LGTM (bar very minor style things) 

I'd quite like to test on both the 1d and Nd case, but extra tests can always be added later :)

Good work!

nickrobinson251 commented on test/name_core.jl:
```diff
@@ -0,0 +1,56 @@
+        @test dim((:x, :y), :y)==2
+        @test dim((:x, :y), :z)==0  # not found
+    end
+    @testset "large case that" begin
```
```suggestion
    @testset "large case" begin
```
